openssl-dynamic/src/main/c/sslcontext.c

```diff
@@ -2331,6 +2341,125 @@ TCN_IMPLEMENT_CALL(void, SSLContext, setPrivateKeyMethod)(TCN_STDARGS, jlong ctx
 #endif // OPENSSL_IS_BORINGSSL
 }
 
+static int tcn_new_session_cb(SSL *ssl, SSL_SESSION *session) {
+    JNIEnv *e = NULL;
+    jboolean result = JNI_FALSE;
+    tcn_ssl_ctxt_t *c = NULL;
+
+    TCN_GET_SSL_CTX(ssl, c);
+    TCN_ASSERT(c != NULL);
+
+    if (tcn_get_java_env(&e) != JNI_OK) {
+        return 0;
+    }
+    if (c->ssl_session_cache == NULL) {
+        return 0;
+    }
+
+    result = (*e)->CallBooleanMethod(e, c->ssl_session_cache, c->ssl_session_cache_creation_method, P2J(ssl), P2J(session));
+
+    if ((*e)->ExceptionCheck(e)) {
+        return 0;
+    }
+
+    if (result == JNI_TRUE) {
+        return 1;
+    }
+    return 0;
+}
+
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+static SSL_SESSION* tcn_get_session_cb(SSL *ssl, const unsigned char *session_id, int len, int *copy) {
+#else
+// Older versions of OpenSSL expect another signature then newer versions
+// See https://github.com/openssl/openssl/blob/OpenSSL_1_0_2/ssl/ssl.h
+static SSL_SESSION* tcn_get_session_cb(SSL *ssl, unsigned char *session_id, int len, int *copy) {
+#endif // OPENSSL_VERSION_NUMBER >= 0x10100000L
+    JNIEnv *e = NULL;
+    jlong result = JNI_FALSE;
+    tcn_ssl_ctxt_t *c = NULL;
+    jbyteArray bArray = NULL;
+
+    TCN_GET_SSL_CTX(ssl, c);
+    TCN_ASSERT(c != NULL);
+
+    if (tcn_get_java_env(&e) != JNI_OK) {
+        return NULL;
+    }
+    if (c->ssl_session_cache == NULL) {
+        return NULL;
+    }
+
+    if ((bArray = (*e)->NewByteArray(e, len)) == NULL) {
```

Comment:
Sorry for a possibly dumb question: Jumping from Java to native will automatically clean up local references when returning to the JVM IIRC.  However, this function appears to be invoked as an SSL callback, so it isn't obvious that this ref will be cleaned up.  Do you need to explicitly free this array on exit?

Reply:
This will still work as all callbacks are invoked from java threads itself.